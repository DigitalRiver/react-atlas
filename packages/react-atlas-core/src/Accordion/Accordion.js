import React from "react";
import PropTypes from "prop-types";
import cx from "classnames";

/**
 * Accordion component creates a tab structure featuring a header and panel.  The MVP implementation requires the use of a div as the accordion child with a title prop
 * @examples <Accordion><div title="accordion title">Some text</div><div title="accordion title 2">Some more text</div></Accordion>
 *
 */
class Accordion extends React.PureComponent {
  constructor(props) {
    super(props);

    // array used to set the activeChildArray value for each child for expanding purposes
    let childArray = [];

    // loop through children and set childArray depending on if expanded prop was set on child.  If not set this defaults to false
    React.Children.map(this.props.children, child => {
      if (child.props.expanded) {
        childArray.push(true);
      } else {
        childArray.push(false);
      }
    });

    // component state
    this.state = {
      "title": props.title,
      "activeChildArray": childArray,
      "expandAll": this.props.expandAll || false,
      "expand": false,
      "multiOpen": this.props.multiOpen || false
    };
  }

  // click handler for header.  Sets activeChildArray value accordingly to expand and collapse the panels
  _click = idx => {
    let newChildArray = [];
    for (let i = 0; i < this.state.activeChildArray.length; i++) {
      if (i === idx) {
        newChildArray.push(!this.state.activeChildArray[i]);
      } else if (!this.state.multiOpen) {
        newChildArray.push(false);
      } else {
        newChildArray.push(this.state.activeChildArray[i]);
      }
    }
    this.setState({
      "activeChildArray": newChildArray
    });
  };

  // expandAll function to expand all panels.  Sets value of all to true in activeChildArray
  _expandAll = () => {
    let newChildArray = [];
    for (let i = 0; i < this.state.activeChildArray.length; i++) {
      newChildArray.push(!this.state.expand);
    }
    this.setState({
      "activeChildArray": newChildArray,
      "expand": !this.state.expand
    });
  };

  render() {
    let { className, children, width, titlePosition, style } = this.props;

    // activeClass is added to activ
    let stateClasses, styleClasses, headerClasses, panelClasses;

    styleClasses = cx({
      "leftAlign": titlePosition !== "right" && titlePosition !== "center",
      "rightAlign": titlePosition === "right",
      "centerAlign": titlePosition === "center"
    });

    // A list of children of the Accordion component
    const accordion_panels = React.Children.map(children, (child, i) => {
      if (this.state.activeChildArray[i] === true && !this.props.disabled) {
        stateClasses = "active";
      } else if (this.props.disabled) {
        stateClasses = "disabled";
      } else {
        stateClasses = "inactive";
      }

      headerClasses = cx("accordion_header", stateClasses, styleClasses);
      panelClasses = cx("accordion_panel", stateClasses);

      let accordion_panel = 
        <div>
          <div
            styleName={headerClasses}
            onClick={() => {
              this._click(i);
            }}
          >
            {child.props.title}
          </div>
          <div styleName={panelClasses}>
            <div styleName={"panel"}>{child}</div>
          </div>
        </div>
      ;
      return accordion_panel;
    });

    return (
      <div className={cx(className)} style={style}>
        {this.state.expandAll && !this.props.disabled ? 
          <div
            styleName={"expandAll"}
            onClick={() => {
              this._expandAll();
            }}
          >
            Expand All
          </div>
         : null}
        <div styleName={"accordion"} style={{ "width": width }}>
          {accordion_panels}
        </div>
      </div>
    );
  }
}

Accordion.propTypes = {
  /**
   * Children should be either a string, an icon/glyphicon, or an image tag.
   * @examples "SomeName", <Accordion>{child}{child}</Accordion>
   */
  "children": PropTypes.node,
  /** An object, array, or string of CSS classes to apply to Accordion.*/
  "className": PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.object,
    PropTypes.array
  ]),
  /**
   * When true, Accordion component will be disabled.
   */
  "disabled": PropTypes.bool,
  /**
   * When true, Accordion will display an "expand all" link that will open all panels.
   * @examples <Accordion expandAll={true}>{children}</Accordion>
   */
  "expandAll": PropTypes.bool,
  /**
   * Accordion will use the expanded prop from each child to determine if the child will be expanded on load.
   * @examples <Accordion><div>value 1</div><div expanded>value 2</div></Accordion>
   */
  "expanded": PropTypes.bool,
  /**
   * When true, Accordion will allow multiple open panels.
   * @examples <Accordion multiOpen={true}>{children}</Accordion>
   */
  "multiOpen": PropTypes.bool,
  /**
   * Pass inline styles here.
   */
  "style": PropTypes.object,
  /**
   * Accordion will use title prop from each child as the title for that child's header.
   * @examples <Accordion><div title={title 1}>value 1</div><div title={title 2}>value 2</div></Accordion>
   */
  "title": PropTypes.string,
  /**
   * Will set the Accordion's title text position left, right, or center.
   * @examples <Accordion titlePosition={left}>{children}</Accordion>
   */
  "titlePosition": PropTypes.string,
  /**
   * Will set the width of the Accordion.
   */
  "width": PropTypes.oneOfType([PropTypes.string, PropTypes.number])
};

Accordion.defaultProps = {
  "width": "100%",
  "disabled": false
};

export default Accordion;
