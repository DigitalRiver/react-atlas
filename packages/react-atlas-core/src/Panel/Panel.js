import React from "react";
import PropTypes from "prop-types";
import cx from "classnames";

class Panel extends React.PureComponent {
  constructor(props) {
    super(props);
  }
  render() {
    const { className, children, style } = this.props;
    return (
      <div style={style} className={cx(className)}>
        {children}
      </div>
    );
  }
}

Panel.propTypes = {
  /** An Object, array, or string of CSS classes to apply to Panel.*/
  "className": PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.object,
    PropTypes.array
  ]),
  /**
     * Text to be displayed can be passed as a child.
     * @examples '<Panel>This is a text hint</Panel>'
     */
  "children": PropTypes.node,

  /**
   * Pass inline styling here. 
   */
  "style": PropTypes.object
};

export default Panel;
