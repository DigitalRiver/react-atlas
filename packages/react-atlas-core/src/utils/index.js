import events from "./events";
import prefixer from "./prefixer";
// import time from './time';
import utils from "./utils";
import classNames from "./classNames";

export { events, prefixer, utils, classNames };
