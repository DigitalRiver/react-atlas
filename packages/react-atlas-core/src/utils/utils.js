export default {
  angleFromPositions(cx, cy, ex, ey) {
    const theta = Math.atan2(ey - cy, ex - cx) + Math.PI / 2;
    return theta * 180 / Math.PI;
  },
  angle360FromPositions(cx, cy, ex, ey) {
    const angle = this.angleFromPositions(cx, cy, ex, ey);
    return angle < 0 ? 360 + angle : angle;
  },
  range(start = 0, stop = null, step = 1) {
    let [_start, _stop] = [0, start];
    if (stop !== null) {
      [_start, _stop] = [start, stop];
    }
    const length = Math.max(Math.ceil((_stop - _start) / step), 0);
    const range = Array(length);

    for (let idx = 0; idx < length; idx++, _start += step) {
      range[idx] = _start;
    }

    return range;
  },
  round(number, decimals) {
    if (!isNaN(parseFloat(number)) && isFinite(number)) {
      const decimalPower = Math.pow(10, decimals);
      return Math.round(parseFloat(number) * decimalPower) / decimalPower;
    }
    return NaN;
  },
  getViewport() {
    return {
      "height": window.innerHeight || document.documentElement.offsetHeight,
      "width": window.innerWidth || document.documentElement.offsetWidth
    };
  },
  cloneObject(object) {
    return JSON.parse(JSON.stringify(object));
  },
  inputTypeForPrototype(prototype) {
    if (prototype === Date) {
      return "date";
    }
    if (prototype === Number) {
      return "number";
    }
    if (prototype === Boolean) {
      return "checkbox";
    }
    return "text";
  },
  prepareValueForInput(value, type) {
    if (type === "date") {
      return new Date(value).toISOString().slice(0, 10);
    }
    if (type === "checkbox") {
      return value ? "on" : null;
    }
    return value;
  },
  getSelection(elem) {
    let start, end, rangeElem, clone;

    if (elem.selectionStart !== undefined) {
      start = elem.selectionStart;
      end = elem.selectionEnd;
    }
    else {
      try {
        elem.focus();
        rangeElem = elem.createTextRange();
        clone = rangeElem.duplicate();

        rangeElem.moveToBookmark(document.selection.createRange().getBookmark());
        clone.setEndPoint('EndToStart', rangeElem);

        start = clone.text.length;
        end = start + rangeElem.text.length;
      }
      catch (e) { /* not focused or not visible */ }
    }

    return { start, end };
  },
  setSelection(elem, selection) {
    let rangeElem;

    try {
      if (elem.selectionStart !== undefined) {
        elem.focus();
        elem.setSelectionRange(selection.start, selection.end);
      }
      else {
        elem.focus();
        rangeElem = elem.createTextRange();
        rangeElem.collapse(true);
        rangeElem.moveStart('character', selection.start);
        rangeElem.moveEnd('character', selection.end - selection.start);
        rangeElem.select();
      }
    }
    catch (e) { /* not focused or not visible */ }
  }
};
