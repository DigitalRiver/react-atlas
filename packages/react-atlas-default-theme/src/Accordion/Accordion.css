.accordion {
    composes: default-border from '../styles.css';
    composes: border-bottom-none from '../styles.css';
    composes: bg-white from '../styles.css';
    composes: default-text from '../styles.css';
    cursor: default;
}
.accordion .disabled {
    opacity: 0.6;
}
.rightAlign {
    text-align: right;
    composes: pad-h-2 from '../styles.css';
}
.leftAlign {
    text-align: left;
    composes: pad-h-2 from '../styles.css';
}
.centerAlign {
    text-align: right;
}
.accordion_header {
    composes: marg-0  from '../styles.css';
    composes: cursor-pointer from '../styles.css';
    composes: bg-silver from '../styles.css';
    composes: default-border-bottom from '../styles.css';
    composes: border-top-none from '../styles.css';
    composes: border-left-none from '../styles.css';
    composes: border-right-none from '../styles.css';
    composes: header-font from '../styles.css';
    line-height: 1.5rem;
    vertical-align: middle;
}
.accordion_header:hover {
    background-color: #bebebe;
    color: white;
}
.accordion_header.active {
    background-color: #00a7e1;
    color: white;
}
.accordion_header.active:hover {
    background-color: #00a7e1;
}
.accordion_header.inactive {

}
.accordion_header.disabled:hover {
    background-color: #e6e6e6;
    color: #333;
    cursor: default;
}

.accordion_panel {
    composes: default-text from '../styles.css';
    composes: default-font from '../styles.css';
}
.accordion_panel div {
    display: none;
    padding-left: 0rem;
    padding-right: 0rem;
}
.accordion_panel.active {
    border-top: none;
    border-left: none;
    border-right: none;
    border-bottom: 1px solid #999;
    max-height: 100%;
    opacity: 1;
    padding: 1rem 0rem 1rem 0rem;
    -moz-transition: .25s;
    -ms-transition: .25s;
    -o-transition: .25s;
    -webkit-transition: .25s;
    transition: .25s;
    transition-timing-function: ease-in;
}
.accordion_panel.active div {
    display: block;
    padding-left: 1rem;
    padding-right: 1rem;
}

.accordion_panel.inactive {
    max-height: 0;
    padding: 0rem;
    opacity: 0;
    -moz-transition: .1s;
    -ms-transition: .1s;
    -o-transition: .1s;
    -webkit-transition: .1s;
    transition: .1s;
}

.expandAll {
    composes: marg-b-1 from '../styles.css';
    composes: primary-text from '../styles.css';
    composes: default-font from '../styles.css';
    cursor: pointer;
}
.expandAll:hover {
    text-decoration: underline;
}