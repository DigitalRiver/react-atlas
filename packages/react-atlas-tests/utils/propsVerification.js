export function verifyPropsDefaultValue(component, expectedProps) {
  //Compares each property value of the component with the expected value

  var result = true;
  var partialResult;

  for (var [compPropName, compPropValue] of expectedProps.entries()) {
    partialResult = component.props()[compPropName] == compPropValue;

    if (!partialResult) {
      console.log(
        "[" +
          compPropName +
          "] -> Got: " +
          component.props()[compPropName] +
          " - expected: " +
          compPropValue
      );
      result = false;
    }
  }
  return result;
}
