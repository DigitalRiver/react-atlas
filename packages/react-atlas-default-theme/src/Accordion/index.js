import * as AccordionStyle from "./Accordion.css";
export { AccordionStyle };
