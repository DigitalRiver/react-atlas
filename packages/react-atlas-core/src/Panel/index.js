import { default as PanelCore } from "./Panel";
export { PanelCore };
