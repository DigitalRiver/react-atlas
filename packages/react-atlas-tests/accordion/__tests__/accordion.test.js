import React from "react";
import { mount } from "enzyme";
import { render } from "enzyme";

import { AccordionCore } from "../../../react-atlas-core/src/Accordion/index";
import { PanelCore as Panel } from "../../../react-atlas-core/src/Panel/index";

import { verifyPropsDefaultValue } from "../../utils/propsVerification";

import renderer from 'react-test-renderer';

describe("Test correct render", () => {

  it("Test correct render", function() {
	const accordion =  <AccordionCore>
						 <Panel title="First">Text for first accordion item</Panel>
                         <Panel title="Second">Text for second accordion item</Panel>
                       </AccordionCore>;
	const tree = renderer.create(accordion).toJSON();
	expect(tree).toMatchSnapshot();
  });
  
  it("Test all collapsed with expand", () => {
	const acc =  mount(<AccordionCore >
						  <Panel>Text for first accordion item</Panel>
						  <Panel>Text for second accordion item</Panel>
						  <Panel>Text for third accordion item</Panel>
						  <Panel>Text for fourth accordion item</Panel>
					   </AccordionCore>);
  });
  
  it("Test all collapsed, disabled", () => {
	const acc =  mount(<AccordionCore disabled>
						  <Panel>Text for first accordion item</Panel>
						  <Panel>Text for second accordion item</Panel>
						  <Panel>Text for third accordion item</Panel>
						  <Panel>Text for fourth accordion item</Panel>
					   </AccordionCore>);
  });
  
  it("Test all expanded", () => {
	const acc =  mount(<AccordionCore>
						 <Panel>Text for first accordion item</Panel>
						 <Panel expanded>Text for second accordion item</Panel>
						 <Panel>Text for third accordion item</Panel>
					   </AccordionCore>);
	expect(acc.state().activeChildArray[0]).toEqual(false);
	expect(acc.state().activeChildArray[1]).toEqual(true);  
	expect(acc.state().activeChildArray[2]).toEqual(false);
  });

  it("Test all collapsed", () => {
	const acc =  mount(<AccordionCore>
						<Panel>first</Panel>
						<Panel>second</Panel>
					   </AccordionCore>);
  });  
  
  it("Test expand one with all collapsed", () => {
	const acc =  mount(<AccordionCore>
						<Panel title='first'>first</Panel>
						<Panel title='second'>second</Panel>
					   </AccordionCore>);
	acc.find('div').forEach((n) => n.props().children=='first' ? n.simulate('click') : null);
	expect(acc.state().activeChildArray[0]).toEqual(true);
	expect(acc.state().activeChildArray[1]).toEqual(false);
  });  

  it("Test switch from one expanded to another expanded", () => {
	const acc =  mount(<AccordionCore>
						<Panel title='first'>first</Panel>
						<Panel title='second'>second</Panel>
					   </AccordionCore>);
	acc.find('div').forEach((n) => n.props().children=='first' ? n.simulate('click') : null);
	expect(acc.state().activeChildArray[0]).toEqual(true);
	expect(acc.state().activeChildArray[1]).toEqual(false);
	acc.find('div').forEach((n) => n.props().children=='second' ? n.simulate('click') : null);
	expect(acc.state().activeChildArray[0]).toEqual(false);
	expect(acc.state().activeChildArray[1]).toEqual(true);
  });  

  it("Test switch from one expanded to another expanded II", () => {
	const acc =  mount(<AccordionCore>
						<Panel title='first'>first</Panel>
						<Panel title='second' expanded >second</Panel>
					   </AccordionCore>);
	expect(acc.state().activeChildArray[0]).toEqual(false);
	expect(acc.state().activeChildArray[1]).toEqual(true);
	acc.find('div').forEach((n) => n.props().children=='first' ? n.simulate('click') : null);
	expect(acc.state().activeChildArray[0]).toEqual(true);
	expect(acc.state().activeChildArray[1]).toEqual(false);
  }); 
  
  it("Test expand one, collapse then same (all collapsed)", () => {
	const acc =  mount(<AccordionCore>
						<Panel title='first'>first</Panel>
						<Panel title='second'>second</Panel>
					   </AccordionCore>);
	acc.find('div').forEach((n) => n.props().children=='first' ? n.simulate('click') : null);
	expect(acc.state().activeChildArray[0]).toEqual(true);
	expect(acc.state().activeChildArray[1]).toEqual(false);
	
	acc.find('div').forEach((n) => n.props().children=='first' ? n.simulate('click') : null);
  });  

  it("Test expand two, multiOpen (all collapsed)", () => {
	const acc =  mount(<AccordionCore multiOpen>
						<Panel title='first'>first</Panel>
						<Panel title='second'>second</Panel>
					   </AccordionCore>);
	acc.find('div').forEach((n) => n.props().children=='first' ? n.simulate('click') : null);
	expect(acc.state().activeChildArray[0]).toEqual(true);
	expect(acc.state().activeChildArray[1]).toEqual(false);
	acc.find('div').forEach((n) => n.props().children=='second' ? n.simulate('click') : null);	
	expect(acc.state().activeChildArray[0]).toEqual(true);
	expect(acc.state().activeChildArray[1]).toEqual(true);
	
	acc.find('div').forEach((n) => n.props().children=='first' ? n.simulate('click') : null);	
	expect(acc.state().activeChildArray[0]).toEqual(false);
	expect(acc.state().activeChildArray[1]).toEqual(true);
	acc.find('div').forEach((n) => n.props().children=='second' ? n.simulate('click') : null);	
  });  

  it("Test expand all by clicking expandAll link", () => {
	const acc =  mount(<AccordionCore expandAll>
						  <Panel>Text for first accordion item</Panel>
						  <Panel>Text for second accordion item</Panel>
						  <Panel>Text for third accordion item</Panel>
						  <Panel>Text for fourth accordion item</Panel>
					   </AccordionCore>);
					   
	acc.find('div').forEach((n) => n.props().children=='Expand All' ? n.simulate('click') : null);
	
	expect(acc.state().activeChildArray[0]).toEqual(true);
	expect(acc.state().activeChildArray[1]).toEqual(true);
	expect(acc.state().activeChildArray[2]).toEqual(true);
	expect(acc.state().activeChildArray[3]).toEqual(true);
  });

  
});
