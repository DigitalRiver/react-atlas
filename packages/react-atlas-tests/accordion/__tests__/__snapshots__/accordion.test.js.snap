// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test correct render Test correct render 1`] = `
<div
  className=""
  style={undefined}
>
  <div
    style={
      Object {
        "width": "100%",
      }
    }
    styleName="accordion"
  >
    <div>
      <div
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        styleName="accordion_header header-inactive"
      >
        First
      </div>
      <div
        styleName="inactive"
      />
    </div>
    <div>
      <div
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        styleName="accordion_header header-inactive"
      >
        Second
      </div>
      <div
        styleName="inactive"
      />
    </div>
  </div>
</div>
`;
